File: backend/src/services/gitService.js
const { exec } = require('child_process');
const fs = require('fs');

const gitCommit = (bookPath, message) => {
    return new Promise((resolve) => {
        if (!fs.existsSync(bookPath)) return resolve(false);
        const path = require('path');
        const gitDir = path.join(bookPath, '.git');
        const hasGit = fs.existsSync(gitDir);
        
        const cmd = hasGit 
            ? 'git add . && git commit -m ' + JSON.stringify(message)
            : 'git init && git add . && git commit -m ' + JSON.stringify(message);
            
        exec(cmd, { cwd: bookPath }, (error, stdout, stderr) => {
            if (error) {
                // If there's nothing to commit, resolve successfully
                if (stdout.includes('nothing to commit') || stderr.includes('nothing to commit')) {
                    return resolve(true);
                }
                console.error(`[Git Error] ${error.message}`);
                return resolve(false);
            }
            console.log(`[Git Time Machine] Commit: ${message}`);
            resolve(true);
        });
    });
};

const getFileHistory = (bookPath, absolutePath) => {
    return new Promise((resolve) => {
        if (!fs.existsSync(bookPath)) return resolve([]);
        const path = require('path');
        const relativePath = path.relative(bookPath, absolutePath).replace(/\\/g, '/');
        
        // %h = short hash, %aI = ISO date, %s = commit subject
        const cmd = `git log --pretty=format:"%h|%aI|%s" -- ` + JSON.stringify(relativePath);
        
        exec(cmd, { cwd: bookPath }, (error, stdout) => {
            if (error || !stdout.trim()) return resolve([]);
            const history = stdout.trim().split('\n').map(line => {
                const [hash, date, message] = line.split('|');
                return { hash, date, message };
            });
            resolve(history);
        });
    });
};

const getFileContentAtCommit = (bookPath, absolutePath, hash) => {
    return new Promise((resolve, reject) => {
        if (!fs.existsSync(bookPath)) return reject(new Error("Repo not found"));
        const path = require('path');
        const relativePath = path.relative(bookPath, absolutePath).replace(/\\/g, '/');
        
        const cmd = `git show ${hash}:` + JSON.stringify(relativePath);
        exec(cmd, { cwd: bookPath }, (error, stdout) => {
            if (error) return reject(error);
            resolve(stdout);
        });
    });
};

module.exports = { gitCommit, getFileHistory, getFileContentAtCommit };

File: src-tauri/src/git_service.rs
use std::process::Command;
use std::path::Path;
use serde::Serialize;

#[derive(Serialize)]
pub struct GitCommitHistory {
    hash: String,
    date: String,
    message: String,
}

pub fn git_commit(book_path: &str, message: &str) -> bool {
    let path = Path::new(book_path);
    if !path.exists() {
        return false;
    }
    let git_dir = path.join(".git");
    if !git_dir.exists() {
        let _ = Command::new("git")
            .arg("init")
            .current_dir(book_path)
            .output();
    }
    let _ = Command::new("git")
        .arg("add")
        .arg(".")
        .current_dir(book_path)
        .output();

    let output = Command::new("git")
        .arg("commit")
        .arg("-m")
        .arg(message)
        .current_dir(book_path)
        .output();

    if let Ok(out) = output {
        if out.status.success() {
            return true;
        }
        let stdout = String::from_utf8_lossy(&out.stdout);
        let stderr = String::from_utf8_lossy(&out.stderr);
        if stdout.contains("nothing to commit") || stderr.contains("nothing to commit") {
            return true;
        }
    }
    false
}

#[tauri::command]
pub fn get_file_history(app: tauri::AppHandle, absolute_path: String) -> Result<Vec<GitCommitHistory>, String> {
    let lib = crate::library_service::get_library(app)?;
    let book = lib.iter().find(|b| absolute_path.starts_with(&b.path))
        .ok_or("Book not found")?;
    let book_path = &book.path;

    if !Path::new(book_path).exists() {
        return Ok(vec![]);
    }
    
    let path_diff = match pathdiff::diff_paths(&absolute_path, book_path) {
        Some(p) => p,
        None => return Ok(vec![]),
    };
    
    let output = Command::new("git")
        .arg("log")
        .arg("--pretty=format:%h|%aI|%s")
        .arg("--")
        .arg(path_diff.to_string_lossy().replace("\\", "/"))
        .current_dir(book_path)
        .output()
        .map_err(|e| e.to_string())?;
        
    if !output.status.success() {
        return Ok(vec![]);
    }
    
    let stdout = String::from_utf8_lossy(&output.stdout);
    if stdout.trim().is_empty() {
        return Ok(vec![]);
    }
    
    let mut history = Vec::new();
    for line in stdout.lines() {
        let parts: Vec<&str> = line.splitn(3, '|').collect();
        if parts.len() == 3 {
            history.push(GitCommitHistory {
                hash: parts[0].to_string(),
                date: parts[1].to_string(),
                message: parts[2].to_string(),
            });
        }
    }
    
    Ok(history)
}

#[tauri::command]
pub fn get_file_content_at_commit(app: tauri::AppHandle, absolute_path: String, hash: String) -> Result<String, String> {
    let lib = crate::library_service::get_library(app)?;
    let book = lib.iter().find(|b| absolute_path.starts_with(&b.path))
        .ok_or("Book not found")?;
    let book_path = &book.path;

    if !Path::new(book_path).exists() {
        return Err("Repo not found".to_string());
    }
    let path_diff = pathdiff::diff_paths(&absolute_path, book_path)
        .ok_or("Cannot diff paths")?;
        
    let relative_path = path_diff.to_string_lossy().replace("\\", "/");
    let target = format!("{}:{}", hash, relative_path);
    
    let output = Command::new("git")
        .arg("show")
        .arg(&target)
        .current_dir(book_path)
        .output()
        .map_err(|e| e.to_string())?;
        
    if !output.status.success() {
        return Err(String::from_utf8_lossy(&output.stderr).to_string());
    }
    
    Ok(String::from_utf8_lossy(&output.stdout).to_string())
}
